Bug fix: auto generate timestamps for hm:t join tables

hayesr opened the pull request:
Currently if a hm:t join model was created using the rails generator, the migration will have timestamps with `null: false` by default.  Loading fixtures will fail because of the database constraint. 

Gist with failing example: https://gist.github.com/hayesr/3756183d34a6305d5143

This change adds timestamps to join models if they are not already defined. (should cover when there's no YAML for the model or when there is but the timestamps aren't defined)

hayesr committed:
Auto generate timestamps for hm:t join tables

sgrif commented:
Sorry this slipped through, I've been taking a bit of a break, so AR pull requests are a bit backed up. This doesn't seem like it should need to be specific to join tables. Would it make sense to just always populate the timestamp columns if they're not specified in the YAML?

hayesr commented:
Makes sense to me! Let me try that route.

hayesr commented:
When you suggested populating timestamps for everything I thought maybe there is something I'm not considering besides models with fixture files and inferred join data. Am I?

From what I can gather, models with fixture files always have their timestamp columns filled. I went blame-diving and it seems this has been the case since the illustrious "Foxy Fixtures" of 2007.

The problem occurs when the fixture loader tries to infer join data. So after this change I believe we will be populating missing timestamps for every case.

Now, the documentation for this magically inferred join data is for HABTM. I expected it to work for HM:T, which it does save this case with the timestamps. One can work around this issue by creating a join fixture without the implicit connection in the other two fixture files. However, this still feels like a bug. HM:T ought to work like HABTM.

It would be nice to add the timestamps to all explicit and implicit rows at once. There doesn't seem to be an easy way to do that without a bigger impact to the code.

azul commented:
I've been bitten by this as well in a Rails 4.2 app. I suspected it would be fixed by now but it does not look like.
I would be happy to rebase the branch and resolve the conflicts if there's a chance of getting this in.

@sgrif, @hayesr  Is there a particular reason why this has not been merged yet?

azul commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -743,11 +743,23 @@ def add_join_records(rows, row, association)
+            if join_class && join_class.record_timestamps
+              now = config.default_timezone == :utc ? Time.now.utc : Time.now
+              now = now.to_s(:db)
+              timestamp_column_names.each do |c_name|
```
`timestamp_column_names` will be called on the fixtures class not on the join class. If the two have a different set of timestamp columns this breaks.

hayesr commented:
@azul I think mainly that it drifted from master, but I was also I think I was waiting on some feedback on my last comment. I remembered this a few weeks ago, but I haven't had a chance to make sure it's still a bug. I'll try for this weekend.

hayesr commented:
I rebased, cleaned this up a bit and squashed commits. @azul I think I addressed your concerns, it should only set columns if they're defined in the join class. 

After almost three years this is ready for another look @sgrif !

azul commented:
@hayesr yes, you addressed my concerns. Thanks a lot!

hayesr committed:
Merge branch 'master' into hmt_auto_timestamp

hayesr committed:
Merge branch 'master' into hmt_auto_timestamp

hayesr committed:
Merge branch 'master' into hmt_auto_timestamp

hayesr committed:
Merge branch 'master' into hmt_auto_timestamp

hayesr committed:
Merge branch 'master' into hmt_auto_timestamp

mikecmpbll commented:
this should be fixed. it's a fairly small change, what would we need to do to get this over the line?
